React/forms.md

```diff
@@ -2,24 +2,266 @@
+import  './new_student_form.css';
+
+class NewStudentForm extends Component {
+  constructor() {
+    super();
+  }
+  render() {
+    return (
+      <div>
+        <form className="new-student-form">
+          <div>
+            <label htmlFor="name">Name:</label>
+            <input name="name" />
+          </div>
+          <div>
+            <label htmlFor="email">Email:</label>
+            <input name="email" />
+          </div>
+          <input className="button success" type="submit" value="Add Student" />
+        </form>
+      </div>
+    );
+  }
+}
+
+export default NewStudentForm;
+```
+
+Now we have a React component to render the form.  It functions just like the HTML form, but can be imported into larger React Components.  Right now the DOM is still maintaining the values of the `name` and `email` fields.  However now that we have them in a React component we can modify the form controls to have React monitor the input fields and handling the form submission.
+
+We can do this by adding `name` and `email` to the `NewStudentForm`'s state in the constructor.
+
+```javascript
+//  new_student_form.js
+...
+constructor() {
+  super();
+
+  this.state = {
+    name: '',
+    email: '',
+  }
+}
+```
+
+The `NewStudentForm` component will track the `name` and `email` as part of it's state.
+
+We can then manage the input fields by setting their value to match the `NewStudentForm`'s state and adding an event handler to respond to changes in the input value.
+
+To link changes in the input field to the `NewStudentForm`'s state we can add an event handler.  So when the input field is edited by the user the event handler function is called which updates the state.
+
+```javascript
+  //  new_student_form.js
+  onNameChange = (event) => {
+    console.log(`Name Field updated ${event.target.value}`)
+    this.setState({
+      name: event.target.value,
+    });
+  }
+```
+
+Then add `onChange` and `value` fields to the `input` in `render`.
+
+```javascript
+<input
+  onChange={ this.onNameChange }
+  value={ this.state.name }
+  name="name"
+/>
+```
+
+Now every time the user types into the name input field the `NewStudentForm`'s state is updated.
+
+**Practice**:  Update the e-mail field to have the `NewStudentForm` component manage it's state as well.
+
+**Question** Why would it be useful to track the state of a form field?
+
+
+## Form Validation
+
+By allowing the `NewStudentForm` component track the status of the form fields this will allow us to:
+- Validate form fields on the fly
+- More easily access the form fields when the form is submitted.
+
+We can perform a validation on the email field with a function like this:
+
+```javascript
+emailValid = () => {
+  return this.state.email.match(/\S+@\S+/);
+}
+```
+
+And give the user feedback on validation with:
+
+```javascript
+<input
+  onChange={ this.handleEmailChange }
+  value={ this.state.email }
+  className={ this.emailValid() ? "valid": "invalid" }
```

Comment:
Is this code in the `render()` function?-- we said in the material that you can't do conditional logic in render, so I wanna acknowledge that you can do these one-line ternaries because it's just JS that's evaluated inline

Reply:
I don't personally care either way, just since we said in the React Hello World Exercise (in Important Notes), someone will ask "didn't we say we can't do that?" so we should be able to say "yes we can" and it might help to have a note to talk about it in the lecture maybe?

https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/React/react-hello-world.md#what-is-jsx

Reply:
ok I added a note